src/codegen/compile.ml

```diff
@@ -1714,6 +1731,31 @@ module Tagged = struct
   let _branch_typed_with env ty retty branches =
     branch_with env retty (List.filter (fun (tag,c) -> can_have_tag ty tag) branches)
 
+  let allocation_barrier env get_object = 
+    (if !Flags.gc_strategy = Flags.Incremental then
+      (* performance gain by first checking the GC state *)
+      E.call_import env "rts" "running_gc" ^^
+      G.if0 (
+        get_object ^^
+        E.call_import env "rts" "allocation_barrier"
+      ) G.nop
+    else 
+      G.nop)  
+  
+  let write_with_barrier env =
+    let (set_value, get_value) = new_local env "written_value" in
+    let (set_location, get_location) = new_local env "write_location" in
+    set_value ^^ set_location ^^
+    (* performance gain by first checking the GC state *)
+    E.call_import env "rts" "running_gc" ^^
+    G.if0 (
+      get_location ^^ get_value ^^
+      E.call_import env "rts" "write_with_barrier"
```

Comment:
I wonder if it would be faster to do the check and site in write_with_barrier, avoiding the locals altogether.

Reply:
Unfortunately, also this is a weird optimization, that I believe gave a high performance improvement. I can remeasure the performance difference and hopefully, we could overcome such in future with a better cost model (or some linker redesign).

Reply:
**Update**: I remeasured this optimization: The runtime improvement for the write barrier alone is **>6%**. Together with the allocation barrier, it is a **>10%** improvement.